src/util/misc.rs

```diff
@@ -160,17 +194,37 @@ mod message_signing {
             address: &Address,
             msg_hash: sha256d::Hash
         ) -> Result<bool, secp256k1::Error> {
+            // Mostly ported from: https://github.com/bitcoinjs/bitcoinjs-message/blob/c43430f4c03c292c719e7801e425d887cbdf7464/index.js#L181-L233
             let pubkey = self.recover_pubkey(&secp_ctx, msg_hash)?;
-            Ok(match address.address_type() {
-                Some(AddressType::P2pkh) => {
-                    *address == Address::p2pkh(&pubkey, address.network)
-                }
-                Some(AddressType::P2sh) => false,
-                Some(AddressType::P2wpkh) => false,
-                Some(AddressType::P2wsh) => false,
-                Some(AddressType::P2tr) => false,
-                None => false,
-            })
+
+            // Use .serialize because we always want compact serialization
+            let pubkey_hash = super::hash160(&pubkey.key.serialize().to_vec()).to_vec();
+
+            match self.segwit_type {
+                None => {
+                    let expected = match bech32_decode(address) {
+                        Ok(data) => data,
+                        Err(_) => {
+                            let redeem_hash = segwit_redeem_hash(&pubkey_hash).to_vec();
+                            let base58_check = get_payload_bytes(address);
+                            return Ok(
+                                (pubkey_hash == base58_check) ||
+                                (redeem_hash == base58_check)
+                            );
+                        }
+                    };
+                    Ok(pubkey_hash == expected)
+                },
+                Some(SegwitType::P2shwpkh) => {
+                    let actual = segwit_redeem_hash(&pubkey_hash).to_vec();
+                    let expected = get_payload_bytes(address);
+                    Ok(actual == expected)
+                },
+                Some(_) => {
+                    let expected = bech32_decode(address).unwrap();
```

Comment:
It is the latter.

> can just call unwrap internally instead of returning the error

I would do that, but bech32_decode is meant to be more generic. It shouldn't fail with the given input in this case, but it can possibly fail with an arbitrary address.

I'd rather have failing cases laid out as to what went wrong rather than just assuming an error will never happen.